File: frontend/lib/features/library/models/book.dart
// frontend/lib/features/library/models/book.dart

class Book {
  final int id;
  final String title;
  final String description;
  final String author;
  final String? coverImage;
  final String? fileUrl;
  final String? level;
  final int? subjectId;
  final String? subject;
  final int userId;
  final String userName;
  final bool isPublic;
  final int viewsCount;
  final int likesCount;
  final int commentsCount;
  final double averageRating;
  final DateTime createdAt;
  final DateTime? updatedAt;
  bool isLiked;

  Book({
    required this.id,
    required this.title,
    required this.description,
    required this.author,
    this.coverImage,
    this.fileUrl,
    this.level,
    this.subjectId,
    this.subject,
    required this.userId,
    required this.userName,
    this.isPublic = true,
    this.viewsCount = 0,
    this.likesCount = 0,
    this.commentsCount = 0,
    this.averageRating = 0.0,
    required this.createdAt,
    this.updatedAt,
    this.isLiked = false,
  });

  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      author: json['author'],
      coverImage: json['cover_image'],
      fileUrl: json['file_url'],
      level: json['level'],
      subjectId: json['subject_id'],
      subject: json['subject'],
      userId: json['user_id'],
      userName: json['user_name'],
      isPublic: json['is_public'] ?? true,
      viewsCount: json['views_count'] ?? 0,
      likesCount: json['likes_count'] ?? 0,
      commentsCount: json['comments_count'] ?? 0,
      averageRating: json['average_rating'] ?? 0.0,
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at']) : null,
      isLiked: json['is_liked'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'author': author,
      'cover_image': coverImage,
      'file_url': fileUrl,
      'level': level,
      'subject_id': subjectId,
      'is_public': isPublic,
    };
  }

  Book copyWith({
    int? id,
    String? title,
    String? description,
    String? author,
    String? coverImage,
    String? fileUrl,
    String? level,
    int? subjectId,
    String? subject,
    int? userId,
    String? userName,
    bool? isPublic,
    int? viewsCount,
    int? likesCount,
    int? commentsCount,
    double? averageRating,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isLiked,
  }) {
    return Book(
      id: id ?? this.id,
      title: title ?? this.title,
      description: description ?? this.description,
      author: author ?? this.author,
      coverImage: coverImage ?? this.coverImage,
      fileUrl: fileUrl ?? this.fileUrl,
      level: level ?? this.level,
      subjectId: subjectId ?? this.subjectId,
      subject: subject ?? this.subject,
      userId: userId ?? this.userId,
      userName: userName ?? this.userName,
      isPublic: isPublic ?? this.isPublic,
      viewsCount: viewsCount ?? this.viewsCount,
      likesCount: likesCount ?? this.likesCount,
      commentsCount: commentsCount ?? this.commentsCount,
      averageRating: averageRating ?? this.averageRating,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isLiked: isLiked ?? this.isLiked,
    );
  }
}

File: backend/app/models/book.py
# backend/app/models/book.py

from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, Boolean, Float
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from ..core.database import Base

class Book(Base):
    __tablename__ = "books"
    __table_args__ = {'extend_existing': True}

    id = Column(Integer, primary_key=True, index=True)
    title = Column(String(255), nullable=False, index=True)
    description = Column(Text, nullable=False)
    author = Column(String(255), nullable=False)
    cover_image = Column(String(500), nullable=True)
    file_url = Column(String(500), nullable=True)
    level = Column(String(50), nullable=True)
    subject_id = Column(Integer, ForeignKey("subjects.id"), nullable=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    is_public = Column(Boolean, default=True)
    views_count = Column(Integer, default=0)
    likes_count = Column(Integer, default=0)
    comments_count = Column(Integer, default=0)
    average_rating = Column(Float, default=0.0)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, onupdate=func.now())

    # ✅ Relations
    user = relationship("User", back_populates="books", foreign_keys=[user_id])
    subject = relationship("Subject", back_populates="books", foreign_keys=[subject_id])
    comments = relationship("BookComment", back_populates="book", cascade="all, delete-orphan", foreign_keys="BookComment.book_id")
    likes = relationship("BookLike", back_populates="book", cascade="all, delete-orphan", foreign_keys="BookLike.book_id")

    def to_dict(self):
        return {
            "id": self.id,
            "title": self.title,
            "description": self.description,
            "author": self.author,
            "cover_image": self.cover_image,
            "file_url": self.file_url,
            "level": self.level,
            "subject_id": self.subject_id,
            "subject": self.subject.name if self.subject else None,
            "user_id": self.user_id,
            "user_name": self.user.full_name if self.user else "Anonyme",
            "is_public": self.is_public,
            "views_count": self.views_count,
            "likes_count": self.likes_count,
            "comments_count": self.comments_count,
            "average_rating": self.average_rating,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }